### DISCUSSION TITLE
[Feature] AV1 video transcoding?

### DISCUSSION THREAD

**Discussion by erkexzcx** [immich-app/immich#5062] [Feature Request]
### The feature

Hello. Am I misunderstanding something, or is indeed VP9 has been superseded by AV1 codec? Why does Immich not have an option to transcode to AV1 instead of VP9?

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by mertalev** [comment]
I'm not sure if the video player on mobile supports AV1; there were some open issues about it last I checked. I'm happy to add it if this is no longer an issue.

**Reply by n00mkrad** [reply]
It does support AV1.

I modified my immich container to transcode to AV1 and it plays flawlessly in Chrome, Firefox, and on Android, and probably more that I haven't tested.

**Reply by mertalev** [reply]
Can you confirm that it's actually playing the remote video on mobile? It will prefer using the local video on the phone if it's available.

**Reply by n00mkrad** [reply]
I 100% confirmed it - On desktop I verified by downloading the transcoded video, on mobile I cleared the cache and then played it.

For the record, the source video is on my NAS (so neither on my phone nor on my desktop PC) and I have also verified that the transcoded video is in AV1.

**Reply by n00mkrad** [reply]
As for the transcoding, I did some hacking and intercepted the ffmpeg calls, simply replacing the args to use libsvtav1 as encoder. 10-bit also works fine, which is generally highly recommended to use when encoding AV1, especially since 10-bit decoding is a requirement for AV1 HW decoders anyway, unlike with H264 for example.

GPU acceleration (NVENC, QSV, etc) should also work with AV1 but I currently don't have that set up on my homeserver.

**Reply by mertalev** [reply]
That's exciting to hear! You could make a PR with what you have and we can work on polishing and merging it in.

**Reply by n00mkrad** [reply]
I will look into it, but honestly I have zero typescript experience.

My current workaround is to use a dummy ffmpeg executable that replaces the encoding arguments and then runs the real binary.

I can try to make a PR if I figure out how to compile the project I guess. Or if that fails I could just do some blind coding, in concept it should be very simple - Add the AV1 option with a different set of ffmpeg args.

I'll look into it either way.

**Reply by n00mkrad** [reply]
Oh, also, it would be great if libsvtav1 in your ffmpeg build could be updated to latest (2.0.0), it has major speed improvements, making it even more viable for CPU encoding.
Currently, immich comes with SVT-AV1 1.8.0.

**Reply by mertalev** [reply]
> My current workaround is to use a dummy ffmpeg executable that replaces the encoding arguments and then runs the real binary.

Oh, I thought you had it in the code. Don't worry about it in that case.

> Oh, also, it would be great if libsvtav1 in your ffmpeg build could be updated to latest (2.0.0)

Thanks for the heads up! I think the latest jellyfin-ffmpeg build uses 2.0.0, so that should be easy.

**Reply by n00mkrad** [reply]
Basically, what needs to be done when selecting AV1:

- Use `-c:v libsvtav1`
- Replace max bitrate (e.g. when set to 8000k) `-b:v 8000k` with `-svtav1-params "mbr=8000k"`
- Map the speed presets like this (my personal recommendation, technically you could go below 4):
```
ultrafast = 12
superfast = 11
veryfast = 10
faster = 9
fast = 8
medium = 7 
slow = 6
slower = 5
veryslow = 4
```

**Reply by n00mkrad** [reply]
I managed to compile the project, I'll play around with the code a bit and see if I can get it working natively. Will make a PR if I get that far.

**Comment by n00mkrad** [comment]
There we go

https://github.com/immich-app/immich/pull/8491
